src/api-service/__app__/timer_retention/__init__.py

```diff
@@ -0,0 +1,97 @@
+import datetime
+import logging
+
+import azure.functions as func
+from onefuzztypes.enums import JobState, TaskState
+
+from ..onefuzzlib.events import get_events
+from ..onefuzzlib.jobs import Job
+from ..onefuzzlib.notifications.main import Notification
+from ..onefuzzlib.repro import Repro
+from ..onefuzzlib.tasks.main import Task
+
+RETENTION_POLICY = datetime.timedelta(minutes=(5))
+
+
+def main(mytimer1: func.TimerRequest, dashboard: func.Out[str]) -> None:  # noqa: F841
+
+    now = datetime.datetime.now(tz=datetime.timezone.utc)
+
+    time_retained = now - RETENTION_POLICY
+
+    time_filter = f"Timestamp lt datetime'{time_retained.isoformat()}'"
+
+    # You have to do notification before task,
+    # because editing the upn for tasks will change the timestamp
+    for notification in Notification.search(raw_unchecked_filter=time_filter):
+        logging.info("Retention Timer Notification Search")
+        logging.info(
+            "Found notification %s older than 18 months. Checking related tasks.",
+            notification.notification_id,
+        )
+        container = notification.container
+        timestamp_list = []
+        for task in Task.search():
+            container_str = str(task.config.containers)
+            if container in container_str:
+                # Need to make sure there isn't a task still using the container.
+                if task.state == TaskState.stopped:
+                    timestamp_list.append(task.timestamp)

```

Comment:
While in a previous comment, I mentioned tasks in the range you care about should be part of the initial query, this is additional complexity that isn't needed.

Once a single task is meets the criteria (within the time window and with the specified container), the notification can be ignored.

Reply:
By ignored do you mean deleted?

Reply:
Your process is currently the psuedo-code:
```
for notification in Notifications.search(...):
   timestamps = []
   for task in Task.search():
      if task containers contain notification.container:
         timestamps.append(tiemstamp)
   if timestamps:
       if get_newest_timestamp < window:
           notification.delete()
```

As soon as you've found one task that is newer than the window, you can stop processing the list.   You don't need to check the rest of the tasks at that point.  Consider if the first of 30000 tasks is newer.  Checking all 30000 tasks is wasteful.

There are three optimizations here, two are discussed in another comment.

1. Perform `Task.search` once to check Notifications.
2. Include the date range you care about in the Task.search.  This puts reduces how many tasks you have to search and the data retrieved from Azure Tables.
3. As soon as the first task includes a container, move onto the next notification, since this one can stay.
4. instead of keeping a list of tasks, keep the set of container names.

This would result in something like this:
```
recent_tasks = Task.search(newer_than_18_months)
recent_containers = get_containers(recent_tasks)
for notification in Notification.search(older_than_18_months):
    if notification.container in recent_containers:
        continue
    notification.delete()
```

Reply:
Just pushed a PR that does basically this. Will address logging PRs in future commits.